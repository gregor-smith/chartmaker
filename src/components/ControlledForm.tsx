import React, { FC, FormEvent } from 'react'


type Props = {
    onSubmit: () => void
}


export const ControlledForm: FC<Props> = ({ children, onSubmit }) => {
    function controlledOnSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault()
        onSubmit()
    }

    return (
        <form onSubmit={controlledOnSubmit}>
            {children}
        </form>
    )
}
