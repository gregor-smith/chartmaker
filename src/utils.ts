import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
