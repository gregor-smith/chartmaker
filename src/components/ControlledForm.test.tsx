import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import { ControlledForm } from './ControlledForm.js'

import { RenderContainer, ignore, fireEvent } from '../test-utils/utils.js'


const container = new RenderContainer()


test('renders form', () => {
    render(
        <ControlledForm onSubmit={ignore}>
            Test children
        </ControlledForm>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('submit calls onSubmit prop', () => {
    const mock = jest.fn<void, []>()

    render(
        <ControlledForm onSubmit={mock}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'submit',
            container.element?.querySelector('form')
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
})
