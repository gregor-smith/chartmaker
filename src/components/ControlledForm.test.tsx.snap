// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form 1`] = `
<div
  id="test-render-container"
>
  <form>
    Test children
  </form>
</div>
`;
